packages/core/src/TextInput.js

```diff
@@ -2,63 +2,164 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 
+import AutoSizeTextarea from 'react-textarea-autosize';
+
 import prefixClass from './utils/prefixClass';
-import rowComp, { getTextLayoutProps, ROW_COMP_ALIGN } from './mixins/rowComp';
+import icBEM from './utils/icBEM';
+import rowComp from './mixins/rowComp';
 
-import EditableText from './EditableText';
+import { PureText, TEXT_ALIGN, VERTICAL_ORDER } from './Text';
+import './styles/TextInput.scss';
 
 export const COMPONENT_NAME = prefixClass('text-input');
+const ROOT_BEM = icBEM(COMPONENT_NAME);
+
+export const BEM = {
+    input: ROOT_BEM.element('input'),
+};
+
+export function TextInputBasicRow({ basic, className }) {
+    return (
+        <div className={className}>
+            {basic}
+        </div>
+    );
+}
+
+TextInputBasicRow.propTypes = {
+    basic: PropTypes.node,
+};
+
+TextInputBasicRow.defaultProps = {
+    basic: undefined,
+};
 
 /**
- * <TextInput>
- * ===========
- * The row component holding an editable `<input>` as its main part.
+ * A `<TextInput>` is a specialized `<TextLabel>`, which holds an editable `<input>`
+ * as its main part.
  * All unknown props are expected to be passed into the underlying `<input>`.
  *
- * What's different from other row component is: _basic text_ is not allowed
- * on a `<TextInput>`, since that place is occupied by an `<input>`.
+ * What's different from other row component is:
+ *   1. It doesn't have `basic` nor `aside`. Instead it holds `label` and `value`/`defaultValue`.
+ *   2. It doesn't support status icons, error message nor tags
  *
  * @example
- * ```jsx
- * <TextInput
- *     value="Hello world"
- *     placeholder="(Unset)"
- *     onChange={event => console.log(event.target.value)} />
- * ```
- */
+```jsx
+# Single-line mode
+<TextInput
+    label="Welcome msg"
+    value="Hello world"
+    placeholder="(Unset)"
+    onChange={event => console.log(event.target.value)} />
 
-function TextInput(props, { align }) {
-    const {
-        wrapperProps,
-        // React props
-        className,
-        children, // strip out from component
-        ...editableTextProps
-    } = props;
+# Multi-line mode
+<TextInput
+    multiLine
+    minRows={5}
+    maxRows={10}
+    label="Comments"
+/>
+
+# Custom rendering
+<TextInput
+    label="Pick a color"
+    renderInput={inputProps => (
+        <input type="color" {...inputProps} />
+    )}
+/>
+```
+ */
 
+function TextInput({
+    label,
+    readOnly,
+    disabled,
+    renderInput,
+    // multi-line mode
+    multiLine,
+    minRows,
+    maxRows,
+    // React props
+    className,
+    children,
+    ...inputProps
+}, context) {
     const rootClassName = classNames(className, COMPONENT_NAME);
-    const textLayoutProps = getTextLayoutProps(align, false);
+    const { textProps } = context;
+
+    const input = (() => {
```

Comment:
Could this be extracted as a functional component?

Reply:
Yes it can. But since we're using many props from `<TextInput>` here, the value might not be obvious.
I'll try this in next commit.

Reply:
@garfieldduck Thanks for the advice! I think the results look great, since it makes tests easier when the input part is extracted.

You can compare the changes in a7e7fd2.